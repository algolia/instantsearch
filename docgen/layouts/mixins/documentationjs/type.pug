mixin type(t, relatedTypes)
  if !t
  else if t.type && t.type === 'FunctionType'
    | (
    each p, index in t.params
      +type(p, relatedTypes)
      if index < t.params.length - 1
        |, 
    | ) => 
    if t.result === null
      | undefined
    else
      +type(t.result, relatedTypes)
  else if t.type && t.type === 'TypeApplication'
    +type(t.expression, relatedTypes)
    | &lt;
    each p, index in t.applications
      +type(p, relatedTypes)
      if index < t.applications.length - 1
        |,
    | &gt;
  else if t.type && t.type === 'StringLiteralType'
    | "!{t.value}"
  else if t.type && t.type === 'OptionalType'
    | [
    +type(t.expression, relatedTypes)
    | ]
  else if t.type && t.type === 'RecordType'
    |{
    for field, idxField in t.fields
      +type(field, relatedTypes)
      if idxField < t.fields.length - 1
        |, 
    |}
  else if t.type && t.type === 'FieldType'
    | !{t.key}: 
    +type(t.value, relatedTypes)
  else if t.type && t.type === 'UnionType'
    each p, index in t.elements
      +type(p, relatedTypes)
      if index < t.elements.length - 1
        ||
  else if t.name
    if t.name !== 'Object' && t.name !== 'Array' && t.name[0] === t.name[0].toUpperCase()
      - const symbol = relatedTypes && relatedTypes.find(t2 => t2 && t.name === t2.name) || {};
      - const isExternal = symbol.kind === 'external'
      if symbol.name && isExternal
        - const firstSee = symbol.tags.find(t => t.title === 'see') || {};
        a.typed-link(href=firstSee.description, target='_blank')=t.name
      else if symbol.name
        a.typed-link(href=`${navPath}#struct-${t.name}`)=t.name
    else
      | !{t.name}
