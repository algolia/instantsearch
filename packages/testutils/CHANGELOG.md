# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.0.1 (2022-11-18)

**Note:** Version bump only for package @instantsearch/testutils
