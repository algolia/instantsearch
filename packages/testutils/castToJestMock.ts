/**
 * Assert that a function is actually a mocked function.
 */
export const castToJestMock = <TFunction extends (...args: any[]) => any>(
  func: TFunction
) => func as jest.MockedFunction<typeof func>;
