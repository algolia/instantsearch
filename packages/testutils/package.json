{
  "name": "@instantsearch/testutils",
  "version": "1.0.1",
  "private": true,
  "dependencies": {
    "@types/enzyme": "3.1.15",
    "enzyme": "3.11.0",
    "instantsearch.js": "4.49.2",
    "preact": "^10.10.0"
  }
}
