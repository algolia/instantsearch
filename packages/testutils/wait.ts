/** Promise for a timeout of a certain number of ms */
export const wait = (ms: number) =>
  new Promise((resolve) => setTimeout(resolve, ms));
