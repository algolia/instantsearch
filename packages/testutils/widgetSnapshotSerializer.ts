import type { Widget } from 'instantsearch.js';
import type { IndexWidget } from 'instantsearch.js/es/widgets/index/index';
import { getWidgetAttribute } from 'instantsearch.js/es/lib/utils';
import { createInitOptions } from '@instantsearch/mocks/createWidget';

function getAttribute(widget: Widget | IndexWidget) {
  try {
    return getWidgetAttribute(widget, createInitOptions());
  } catch {
    return undefined;
  }
}

/**
 * Jest serializer for an InstantSearch widget.
 *
 * It looks like this:
 *
 * Widget($$type)
 * Widget($$type) {
 *  $$widgetType: string
 *  attribute: string
 * }
 */
export const widgetSnapshotSerializer: jest.SnapshotSerializerPlugin = {
  serialize(widget: Widget | IndexWidget, { indent }, indentation) {
    const keys = {
      $$widgetType: widget.$$widgetType,
      attribute: getAttribute(widget),
    };

    const widgetName = `Widget(${widget.$$type || 'unknown'})`;

    const content = Object.entries(keys)
      .filter(([_key, value]) => value)
      .map(([key, value]) => `${indentation}${indent}${key}: ${value}`)
      .join('\n');

    if (content) {
      return `${widgetName} {
${content}
${indentation}}`;
    }

    return widgetName;
  },
  test(value) {
    return Boolean(value?.$$type);
  },
};
