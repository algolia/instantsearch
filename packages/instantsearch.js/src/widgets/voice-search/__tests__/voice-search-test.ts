/**
 * @jest-environment jsdom
 */

import {
  createSearchClient,
  createSingleSearchResponse,
} from '@instantsearch/mocks';
import { castToJestMock } from '@instantsearch/testutils/castToJestMock';
import algoliasearchHelper, {
  SearchResults,
  SearchParameters,
} from 'algoliasearch-helper';
import { render as preactRender } from 'preact';

import {
  createInitOptions,
  createRenderOptions,
} from '../../../../test/createWidget';
import voiceSearch from '../voice-search';

import type { VoiceSearchProps } from '../../../components/VoiceSearch/VoiceSearch';
import type { VoiceSearchHelper } from '../../../lib/voiceSearchHelper/types';
import type { Widget } from '../../../types';
import type { VoiceSearchWidgetParams } from '../voice-search';
import type { AlgoliaSearchHelper as Helper } from 'algoliasearch-helper';
import type { VNode } from 'preact';

const render = castToJestMock(preactRender);
jest.mock('preact', () => {
  const module = jest.requireActual('preact');

  module.render = jest.fn();

  return module;
});

type DefaultSetupWrapper = {
  container: HTMLDivElement;
  widget: Widget;
  widgetInit: (helper: Helper) => void;
  widgetRender: (helper: Helper) => void;
};

function defaultSetup(
  opts: Omit<VoiceSearchWidgetParams, 'container'> = {}
): DefaultSetupWrapper {
  const container = document.createElement('div');
  const widget = voiceSearch({ container, ...opts });

  const widgetInit = (helper: Helper): void => {
    if (!widget.init) {
      throw new Error('VoiceSearch widget has no init method.');
    }

    widget.init(
      createInitOptions({
        helper,
        state: helper.state,
      })
    );
  };

  const widgetRender = (helper: Helper): void => {
    if (!widget.render) {
      throw new Error('VoiceSearch widget has no render method.');
    }

    widget.render(
      createRenderOptions({
        helper,
        state: helper.state,
        results: new SearchResults(helper.state, [
          createSingleSearchResponse(),
        ]),
      })
    );
  };

  return { container, widget, widgetInit, widgetRender };
}

describe('voiceSearch()', () => {
  let helper: Helper;

  beforeEach(() => {
    render.mockClear();

    helper = algoliasearchHelper(createSearchClient(), '', {});
    helper.setQuery = jest.fn();
    helper.search = jest.fn();
    helper.state = new SearchParameters({ query: '' });
  });

  describe('Usage', () => {
    it('throws without container', () => {
      expect(() => {
        voiceSearch({
          // @ts-expect-error
          container: undefined,
        });
      }).toThrowErrorMatchingInlineSnapshot(`
"The \`container\` option is required.

See documentation: https://www.algolia.com/doc/api-reference/widgets/voice-search/js/"
`);
    });

    it('creates custom voice helper', () => {
      const voiceHelper: VoiceSearchHelper = {
        isBrowserSupported: () => true,
        dispose: () => {},
        getState: () => ({
          isSpeechFinal: true,
          status: 'askingPermission',
          transcript: '',
        }),
        isListening: () => true,
        startListening: () => {},
        stopListening: () => {},
      };

      const { widgetInit, widget } = defaultSetup({
        createVoiceSearchHelper: () => voiceHelper,
      });

      widgetInit(helper);

      expect((widget as any)._voiceSearchHelper).toBe(voiceHelper);
    });
  });

  describe('Rendering', () => {
    test('renders during init()', () => {
      const { widgetInit } = defaultSetup();
      widgetInit(helper);

      const firstRender = render.mock.calls[0][0] as VNode<VoiceSearchProps>;

      expect(render).toHaveBeenCalledTimes(1);
      expect(firstRender.props).toMatchSnapshot();
    });

    test('renders during render()', () => {
      const { container, widgetInit, widgetRender } = defaultSetup();
      widgetInit(helper);
      widgetRender(helper);

      const firstRender = render.mock.calls[0][0] as VNode<VoiceSearchProps>;
      const secondRender = render.mock.calls[1][0] as VNode<VoiceSearchProps>;
      const firstContainer = render.mock.calls[0][1];
      const secondContainer = render.mock.calls[1][1];

      expect(render).toHaveBeenCalledTimes(2);
      expect(firstRender.props).toMatchSnapshot();
      expect(firstContainer).toEqual(container);
      expect(secondRender.props).toMatchSnapshot();
      expect(secondContainer).toEqual(container);
    });

    test('sets the correct CSS classes', () => {
      const { widgetInit } = defaultSetup();

      widgetInit(helper);

      const firstRender = render.mock.calls[0][0] as VNode<VoiceSearchProps>;

      expect(
        (firstRender.props as VoiceSearchProps).cssClasses
      ).toMatchSnapshot();
    });
  });
});
