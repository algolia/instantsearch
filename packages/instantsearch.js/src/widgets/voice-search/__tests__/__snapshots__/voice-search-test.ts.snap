// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`voiceSearch() Rendering renders during init() 1`] = `
{
  "cssClasses": {
    "button": "ais-VoiceSearch-button",
    "root": "ais-VoiceSearch",
    "status": "ais-VoiceSearch-status",
  },
  "isBrowserSupported": false,
  "isListening": false,
  "templates": {
    "buttonText": [Function],
    "status": [Function],
  },
  "toggleListening": [Function],
  "voiceListeningState": {
    "errorCode": undefined,
    "isSpeechFinal": false,
    "status": "initial",
    "transcript": "",
  },
}
`;

exports[`voiceSearch() Rendering renders during render() 1`] = `
{
  "cssClasses": {
    "button": "ais-VoiceSearch-button",
    "root": "ais-VoiceSearch",
    "status": "ais-VoiceSearch-status",
  },
  "isBrowserSupported": false,
  "isListening": false,
  "templates": {
    "buttonText": [Function],
    "status": [Function],
  },
  "toggleListening": [Function],
  "voiceListeningState": {
    "errorCode": undefined,
    "isSpeechFinal": false,
    "status": "initial",
    "transcript": "",
  },
}
`;

exports[`voiceSearch() Rendering renders during render() 2`] = `
{
  "cssClasses": {
    "button": "ais-VoiceSearch-button",
    "root": "ais-VoiceSearch",
    "status": "ais-VoiceSearch-status",
  },
  "isBrowserSupported": false,
  "isListening": false,
  "templates": {
    "buttonText": [Function],
    "status": [Function],
  },
  "toggleListening": [Function],
  "voiceListeningState": {
    "errorCode": undefined,
    "isSpeechFinal": false,
    "status": "initial",
    "transcript": "",
  },
}
`;

exports[`voiceSearch() Rendering sets the correct CSS classes 1`] = `
{
  "button": "ais-VoiceSearch-button",
  "root": "ais-VoiceSearch",
  "status": "ais-VoiceSearch-status",
}
`;
